### PR TITLE
Allow for appending tasks to co-exist with each other.

### PR DESCRIPTION
**Pull request by imply-cheddar**
### Description

Add a config parameter for appending tasks to allow them to
use a SHARED lock.  This will allow multiple appending tasks
to add segments to the same datasource at the same time.

This config should actually be the default, but it is added
as a config to enable a smooth transition/validation in
production settings before forcing it as the default
behavior going forward.

This change leverages the TaskLockType.SHARED that existed
previously, this used to carry the semantics of a READ lock,
which was "escalated" when the task wanted to actually
persist the segment.  As of many moons before this diff, the
SHARED lock had stopped being used but was still piped into
the code.  It turns out that with a few tweaks, it can be
adjusted to be a shared lock for append tasks to allow them
all to write to the same datasource, so that is what this does.

##### Key changed/added classes in this PR
 * New task context on append tasks: `useSharedLock`.  Defaults to false.  Exists to validate functionality in production environments before shared locks are made the default and only type of lock for appending tasks to use.

This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java [hunk 1]
```diff
@@ -286,10 +286,11 @@ public boolean determineLockGranularityAndTryLock(TaskActionClient client, List<
         Tasks.FORCE_TIME_CHUNK_LOCK_KEY,
         Tasks.DEFAULT_FORCE_TIME_CHUNK_LOCK
     );
+    final boolean useSharedLock = getContextValue(Tasks.USE_SHARED_LOCK, false);
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:289]: Should we also check if `appendingToExisting` is set to true?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java [hunk 2]
```diff
@@ -581,7 +581,7 @@ private boolean isTaskLocksValid(Task task, List<Interval> intervals)
             final List<TaskLockPosse> lockPosses = getOnlyTaskLockPosseContainingInterval(task, interval);
             // Tasks cannot enter the critical section with a shared lock
             return lockPosses.stream().map(TaskLockPosse::getTaskLock).allMatch(
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:583]: > [ERROR] indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:583 -- Stream.allMatch(x -> !(...)) can be replaced with noneMatch(...)

Intellij inspection failed at this line.

**Review by jihoonson — COMMENTED** [review]
The key change LGTM. Append tasks should use the shared lock in timeChunk locking, so that they can run at the same time. Besides my comments, can you check the CI failure? These failures seem legit.

```
[ERROR] org.apache.druid.indexing.overlord.TaskLockboxTest.testDoInCriticalSectionWithSharedLock(org.apache.druid.indexing.overlord.TaskLockboxTest)
[ERROR]   Run 1: TaskLockboxTest.testDoInCriticalSectionWithSharedLock:494
[ERROR] Errors: 
[ERROR] org.apache.druid.indexing.overlord.TaskLockboxTest.testPreemptionAndDoInCriticalSection(org.apache.druid.indexing.overlord.TaskLockboxTest)
[ERROR]   Run 1: TaskLockboxTest.testPreemptionAndDoInCriticalSection:554->tryTimeChunkLock:142 » ISE
[ERROR] org.apache.druid.indexing.overlord.TaskLockboxTest.testTryMixedLocks(org.apache.druid.indexing.overlord.TaskLockboxTest)
[ERROR]   Run 1: TaskLockboxTest.testTryMixedLocks:214->tryTimeChunkLock:142 » ISE Failed to fi...
[ERROR] org.apache.druid.indexing.overlord.TaskLockboxTest.testTrySharedLock(org.apache.druid.indexing.overlord.TaskLockboxTest)
[ERROR]   Run 1: TaskLockboxTest.testTrySharedLock:182->tryTimeChunkLock:142 » IllegalArgument ...
```

**Review by jihoonson — APPROVED** [review]
+1 after CI passes. It makes sense to not document the new context as it's for doing experiments and making the new behavior a default.

**Review by gianm — APPROVED** [review]
I just looked at the API + concept level, but those LGTM

**Comment by jihoonson** [comment]
The last commit only removed an unused import. Merging this PR as the first phase has passed in Travis.

**Pull request merged**